Remove dependency on containers

ChrisPenner opened the pull request:
Removes the dependency on `containers` by implementing `unionWith` manually, addressing https://github.com/kowainik/typerep-map/issues/30

I believe this ends up being O(n + m) unionWith, via the following reasoning:

1. We assume that all incoming TypeRepMaps are already sorted into "cachedBinarySearch order" using `fromSortedList`.
2. We construct the index mapping from "cached" order to normal sorting order using `generateOrderMapping` for both maps; `O(n) +  O(m) == O(n + m)`.
3. We walk the elements in sorted order, pulling out elements using O(1) array indexing using our index mapping, Each step is O(1), as it's just a few array indexings, and a comparison. We build the output structure in sorted order as we recurse using a merge-sort. Overall the merge-sort is O(n + m) since we hit each element at most once. We handle equal elements 2 at a time: `O(n + m)`
4. After we've merged everything, we have a sorted list of triples, we use `fromSortedTriples` to transform back to tree layout `O(n + m)`.

Let me know if I've made any faulty reasoning here 😄 

This also enables #94

ChrisPenner commented on src/Data/TypeRepMap/Internal.hs:
```diff
@@ -266,25 +264,53 @@ hoistWithKey f (TypeRepMap as bs ans ks) = TypeRepMap as bs newAns ks
-                                  (fromTripleList $ toTriples m1)
-                                  (fromTripleList $ toTriples m2)
+unionWith f ma mb = do
+    fromSortedTriples $ mergeMaps orderingM1 orderingM2
```
The result of "mergeMaps" *SHOULD* be sorted, but for some reason switch `fromSortedTriples` -> `fromTriples` does fix some of the failing tests... mergeMaps must be returning things in the wrong order somehow...

ChrisPenner commented on src/Data/TypeRepMap/Internal.hs:
```diff
@@ -464,11 +493,18 @@ fromSortedList l = runST $ do
+-- I.e. fmap (fromSortedList [1, 2, 3, 4, 5] !!) (generateOrderMapping 5) == [1, 2, 3, 4, 5]
+generateOrderMapping :: Int -> [Int]
+generateOrderMapping 0 = []
+generateOrderMapping len = fromSortedList [0..len-1]
```
Something is fishy here

chshersh commented:
Hi @ChrisPenner! Thanks a lot for your work in this PR! Really appreciate that, and the work is amazing :clap: :exploding_head: 
I'm impressed by your contribution! :trophy: 

Sorry for the delay in review, and I'm afraid it will take us some time to review the changes and get back to you with our feedback, as it is a very busy time at the moment :disappointed: But I've added Hacktoberfest labels to your PRs, so they still should count towards your contribution :shirt: :deciduous_tree:

chshersh commented on src/Data/TypeRepMap/Internal.hs:
```diff
@@ -170,7 +169,12 @@ prop> member @a (one (x :: f a)) == True
-one x = insert x empty
+one x = TypeRepMap (primArrayFromListN 1 [fa])
+                   (primArrayFromListN 1 [fb])
+                   (pure v)
+                   (pure k)
```
I think, using type applications to create arrays with `pure` will help here for readability in the future
```suggestion
                   (pure @Array v)
                   (pure @Array k)
```

chshersh commented on src/Data/TypeRepMap/Internal.hs:
```diff
@@ -266,25 +270,32 @@ hoistWithKey f (TypeRepMap as bs ans ks) = TypeRepMap as bs newAns ks
+    mergeMaps as [] = as
+    mergeMaps [] bs = bs
+    -- Merge
+    mergeMaps (a@(af, _, _) : as) (b@(bf, _, _) : bs)
+      -- Fingerprints are equal, union the elements using our function
+      -- If the incoming maps were de-duped, there shouldn't be any other equivalent
+      -- fingerprints
+      | af == bf = combine a b : mergeMaps as bs
+      -- First fingerprint must not be in the second map or we would have seen it by now
+      -- Add it to the result as-is
+      | af < bf = a : mergeMaps as (b : bs)
+      | otherwise = b : mergeMaps (a:as) bs
```
I would replace multiple comparisons with a single call to `compare` function and pattern matching. Otherwise, the merging implementation looks good :+1:

chshersh commented on src/Data/TypeRepMap/Internal.hs:
```diff
@@ -434,10 +455,13 @@ calcFp :: forall a . Typeable a => Fingerprint
 
 fromTriples :: [(Fingerprint, Any, Any)] -> TypeRepMap f
-fromTriples kvs = TypeRepMap (GHC.fromList fpAs) (GHC.fromList fpBs) (GHC.fromList ans) (GHC.fromList ks)
+fromTriples kvs = fromSortedTriples . sortWith fst3 . nubByFst $ kvs
```
I think, you can eta-reduce this function after refactoring

```suggestion
fromTriples = fromSortedTriples . sortWith fst3 . nubByFst
```

chshersh commented on src/Data/TypeRepMap/Internal.hs:
```diff
@@ -464,11 +488,29 @@ fromSortedList l = runST $ do
                 writeArray result i (indexArray origin newFirst)
                 loop (2 * i + 2) (newFirst + 1)
 
+-- Returns a list of indexes which represents the "sorted" order of an array generated by
+-- fromSortedList of the provided length.
+-- I.e. fmap (fromSortedList [1, 2, 3, 4, 5] !!) (generateOrderMapping 5) == [1, 2, 3, 4, 5]
```
I think, it also will be useful to give a few documentation examples of the `generateOrderMapping` function: what it produces for different outputs.

Also, it looks like you can implement a property-based test for this function using `hedgehog`. Something like:

```haskell
n <- forAll genInt
fmap (fromSortedList [1 .. n] !!) (generateOrderMapping n) === [1 .. n]
```

Never hurts to have more tests about invariants :slightly_smiling_face:

chshersh commented on src/Data/TypeRepMap/Internal.hs:
```diff
@@ -266,25 +270,32 @@ hoistWithKey f (TypeRepMap as bs ans ks) = TypeRepMap as bs newAns ks
-                  $ Map.unionWith combine
-                                  (fromTripleList $ toTriples m1)
-                                  (fromTripleList $ toTriples m2)
+unionWith f ma mb = do
```
You wrote an excellent description of how `unionWith` works! Could you, please put into the Haddock documentation for this function as well? E.g. under the _Implementation description_ note or something like that. And the performance analysis is also quite useful :slightly_smiling_face:

chshersh commented on src/Data/TypeRepMap/Internal.hs:
```diff
@@ -464,11 +488,29 @@ fromSortedList l = runST $ do
+    orderMappingArr <- newPrimArray len
+    _ <- loop orderMappingArr 0 0
+    primArrayToList <$> unsafeFreezePrimArray orderMappingArr
+    where
```
Just a minor stylistic note, but `where` is indented with two spaces according to our style guide.
```suggestion
  where
```

chshersh reviewed (commented):
@ChrisPenner Again, amazing work! Thank's a lot for working on this :1st_place_medal: 

I left some comments, but they are mostly about code style and documentation.

ChrisPenner commented:
I've addressed all the feedback if you'd like to take another look 😄 

I left the nitty gritty bits out of the docstring, but added it in the implementation and added a few Big O notations to the new docstrings; also added the property test as suggested. Cheers 😄

vrom911 reviewed (approved):
@ChrisPenner thank you so much for such great work! Really neat implementation 👍🏼

chshersh commented:
@ChrisPenner Let us know what is the next PR that logically continues this awesome chain of improvements :slightly_smiling_face: So you can rebase it, and we can review it as well.

ChrisPenner commented:
@vrom911 @chshersh The next three are independent, I rebased them all, you can merge in any order 😄 ; might be a merge conflict or two in the property tests, but just because all the new tests are added in the same spot 😋
